File: backend/app/schemas/traceability.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel

from app.db.models import ArtifactType, DocumentStatus, DocumentType, InputType, LinkType


class TraceabilityLinkCreate(BaseModel):
    source_type: ArtifactType
    source_id: uuid.UUID
    target_type: ArtifactType
    target_id: uuid.UUID
    link_type: LinkType


class TraceabilityLinkResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    source_type: ArtifactType
    source_id: uuid.UUID
    target_type: ArtifactType
    target_id: uuid.UUID
    link_type: LinkType
    created_at: datetime

    model_config = {"from_attributes": True}


class RequirementInputSummary(BaseModel):
    id: uuid.UUID
    title: Optional[str]
    input_type: InputType
    created_at: datetime

    model_config = {"from_attributes": True}


class DocumentCoverage(BaseModel):
    id: uuid.UUID
    document_type: DocumentType
    title: str
    status: DocumentStatus
    created_at: datetime

    model_config = {"from_attributes": True}


class CoverageStats(BaseModel):
    total_requirement_inputs: int
    total_documents: int
    linked_requirement_inputs: int
    document_types_present: list[DocumentType]
    document_types_missing: list[DocumentType]
    coverage_pct: float


class TraceabilityMatrix(BaseModel):
    project_id: uuid.UUID
    requirement_inputs: list[RequirementInputSummary]
    documents: list[DocumentCoverage]
    links: list[TraceabilityLinkResponse]
    coverage: CoverageStats


class AutoLinkResult(BaseModel):
    links_created: int
    links_skipped: int
    message: str

File: frontend/src/types/traceability.ts
import type { DocumentType } from "./document";

export type ArtifactType = "requirement_input" | "document" | "task" | "pipeline_step";
export type LinkType = "derived_from" | "implements" | "tests" | "conflicts_with";

export interface TraceabilityLink {
  id: string;
  project_id: string;
  source_type: ArtifactType;
  source_id: string;
  target_type: ArtifactType;
  target_id: string;
  link_type: LinkType;
  created_at: string;
}

export interface RequirementInputSummary {
  id: string;
  title: string | null;
  input_type: string;
  created_at: string;
}

export interface DocumentCoverage {
  id: string;
  document_type: DocumentType;
  title: string;
  status: string;
  created_at: string;
}

export interface CoverageStats {
  total_requirement_inputs: number;
  total_documents: number;
  linked_requirement_inputs: number;
  document_types_present: DocumentType[];
  document_types_missing: DocumentType[];
  coverage_pct: number;
}

export interface TraceabilityMatrix {
  project_id: string;
  requirement_inputs: RequirementInputSummary[];
  documents: DocumentCoverage[];
  links: TraceabilityLink[];
  coverage: CoverageStats;
}

export interface AutoLinkResult {
  links_created: number;
  links_skipped: number;
  message: string;
}
